src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,811 @@
+  void Verify() const {
+    CHECK(upper_factor == kPosInf || upper_factor % lower_factor == 0);
+  }
+
+  Expr NormalizeWithScale(int64_t sscale) const {
+    Expr res = this->index;
+    Type dtype = this->type;
+    CHECK_EQ(this->type, dtype);
+    if (this->scale == 0) {
+      return make_const(dtype, 0);
+    }
+    if (this->upper_factor != SplitExprNode::kPosInf) {
+      res = res % make_const(dtype, this->upper_factor);
+    }
+    if (this->lower_factor != 1) {
+      res = res / make_const(dtype, this->lower_factor);
+    }
+    sscale *= this->scale;
+    if (sscale != 1) {
+      CHECK(!dtype.is_uint() || sscale > 0);
+      res = res * make_const(dtype, sscale);
+    }
+    return res;
+  }
+
+  Expr Normalize() const final {
+    return NormalizeWithScale(1);
+  }
+
+  void MulToSelf(int64_t scale) {
+    this->scale *= scale;
+  }
+
+  inline bool IndexEqual(const SplitExpr& other) const;
+
+  /*! \brief positive infty */
+  static const constexpr int64_t kPosInf = ConstIntBoundNode::kPosInf;
+  static constexpr const char* _type_key = "arith.SplitExpr";
+  TVM_DECLARE_NODE_TYPE_INFO(SplitExprNode, CanonicalExprNode);
+};
+
+TVM_DEFINE_COW_NODE_REF(SplitExpr, Expr, SplitExprNode);
+
+inline bool SplitExprNode::IndexEqual(const SplitExpr& other) const {
+  if (index.same_as(other->index)) return true;
+  return ir::Equal(index, other->index);
+}
+
+/*!
+ * \brief Normal form that represent sum of expressions.
+ *
+ *  result = sum(args) + base.
+ */
+class SumExprNode : public CanonicalExprNode {
+ public:
+  /*!
+   * \brief arguments to be summed up.
+   *
+   * args are divided into segments with the same index.
+   * within each segment, the SplitExpr is ordered in descending order of lower_factor.
+   *
+   * \note Can be mutated by TryMergeSplitExpr, which is idempotent
+   */
+  mutable std::vector<SplitExpr> args;
+  /*! \brief Base value in the summation. */
+  int64_t base{0};
+
+  /*! \brief Simplify the args by merging SplitExprs */
+  void SimplifySplitExprs() const {
+    // NOTE: This algorithm relies on the factor that args are divided into segments
+    // and each segment is sorted in descending order of lower_factor.
+    for (size_t i = 0; i < args.size(); ++i) {
+      if (args[i]->scale == 0) continue;
+      for (size_t j = i + 1; j < args.size(); ++j) {
+        SplitExpr& lhs = args[i];
+        SplitExpr& rhs = args[j];
+        if (!lhs->IndexEqual(rhs)) break;
+        if (lhs->upper_factor < rhs->lower_factor) break;
+        if (lhs->lower_factor == rhs->upper_factor &&
+            lhs->scale % rhs->scale == 0 &&
+            lhs->lower_factor == (lhs->scale / rhs->scale) * rhs->lower_factor) {
+          // Merge condition:
+          // - x = lhs->index % lhs->upper_factor
+          // - s = lhs->scale / rhs->scale
+          // - c = rhs->lower_factor
+          //
+          //    ((x / (c * s)) * s + (x % (c * s)) / c
+          // => ((x / c) / s) * s + ((x / c) % s)
+          // => (x / c)
```

Comment:
Honestly speaking, I can't understand this algorithm. Probably I have to return to it in a better
state of mind. Expanding the explanation may be helpful too.

Reply:
The simplification rule and proof are correct.  It's based on two basic rules:

```c++
Rule 1:  (x % (c * s)) / c  =  (x / c) % s
Proof:
   x can always be decomposed into p * c * s + q * c + r  where  0 <= q * c + r < c * s  and  0 <= r  <  c.
   Then, lhs = ((p * c * s + q * c + r) % (c * s)) / c = (q * c + r) / c = q
         rhs = ((p * c * s + q * c + r) / c) % s = (p * s + q) % s = q
   Thus, lhs = rhs

Rule 2:  (x / s) * s + x % s = x
```

Reply:
The rule works for both trunc div and floor version. Mainly because that the first rule only involves mul div and mod. And you can simply take abs of all operands and then take addd the final sign. The second rule is an invariant for both types of div